### PR TITLE
Bug 1754070: Allow setting root volume via installconfig

### PR DESCRIPTION
**Pull request by mandre**
The `rootVolume` setting can now be passed via installconfig's
machinepool struct to let users use a more performant volume type than
the default ephemeral disk.

### CODE REVIEW

File: pkg/tfvars/openstack/openstack.go [hunk 1]
```diff
@@ -18,10 +18,19 @@ type config struct {
 	IngressVIP      string `json:"openstack_ingress_ip,omitempty"`
 	TrunkSupport    string `json:"openstack_trunk_support,omitempty"`
 	OctaviaSupport  string `json:"openstack_octavia_support,omitempty"`
+	Size            int    `json:"openstack_master_root_volume_size,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/tfvars/openstack/openstack.go]: nit: `RootVolumeSize` and `RootVolumeType`

File: pkg/asset/machines/openstack/machines.go [hunk 2]
```diff
@@ -86,10 +86,10 @@ func provider(clusterID string, platform *openstack.Platform, mpool *openstack.M
 			Kind:       "OpenstackProviderSpec",
 		},
 		Flavor: mpool.FlavorName,
-		/*RootVolume: openstackprovider.RootVolume{
-			VolumeType: pointer.StringPtr(mpool.Type),
-			Size:       pointer.Int64Ptr(int64(mpool.Size)),
-		},*/
+		RootVolume: &openstackprovider.RootVolume{
+			Size:       mpool.RootVolume.Size,
+			SourceType: mpool.RootVolume.Type,
```
Comment (mandre) [inline pkg/asset/machines/openstack/machines.go]: I doubt that `SourceType` is correct here.
Looking at https://github.com/gophercloud/gophercloud/blob/8f47ec3/openstack/compute/v2/extensions/bootfromvolume/requests.go#L81, we'd probably want `VolumeType` instead, however that doesn't seem implemented in CAPO?

https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/7e4e4a8/api/v1alpha2/types.go#L112

Comment (Fedosin) [reply pkg/asset/machines/openstack/machines.go]: Source type should be "image", DeviceType is mpool.RootVolume.Type

**Comment by eparis** [comment]
abhinav or one of the install team need to review one final time.

File: docs/user/openstack/README.md [hunk 3]
```diff
@@ -93,7 +93,7 @@ openstack object store account set --property Temp-URL-Key=superkey
 
 ### Disk Requirements
 
-Etcd is run on your control plane nodes, and it has disk requirements that need to be met to ensure the stability of your cluster. It is important that you make sure that the ephemeral disk that gets attached to instances of your chosen flavor meets [these requirements](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/hardware.md#disks). If this is not the case, then please refer to the known issue documentation [here](known-issues.md#boot-from-volume-support).
+Etcd runs on the control plane nodes, and has disk requirements that need to be met to ensure the stability of the cluster. If the ephemeral disk that gets attached to instances of the chosen flavor does not meet [etcd requirements](https://github.com/etcd-io/etcd/blob/master/Documentation/op-guide/hardware.md#disks), check if the cloud has a more performant volume type and use a [custom `install-config.yaml`](customization.md) to deploy the cluster with root volumes.
```
Comment (wking) [inline docs/user/openstack/README.md]: Nit: "the cluster with root volumes" -> "the control plane with root volumes".  Compute nodes don't come into this.

File: pkg/types/openstack/machinepool.go [hunk 4]
```diff
@@ -6,6 +6,9 @@ type MachinePool struct {
 	// FlavorName defines the OpenStack Nova flavor.
 	// eg. m1.large
 	FlavorName string `json:"type"`
+
+	// RootVolume defines the root volume for instances in the machine pool.
+	RootVolume `json:"rootVolume"`
```
Comment (wking) [inline pkg/types/openstack/machinepool.go]: I think this should be

    RootVolume *RootVolume `json:"rootVolume,omitempty"`

so we don't render empty structs when serializing.

Comment (mandre) [reply pkg/types/openstack/machinepool.go]: We would then have to check if the stuct is present in https://github.com/openshift/installer/pull/2394/files#diff-a2bd450b1cd3199485e614928c597a91R90, wouldn't we?

Comment (wking) [reply pkg/types/openstack/machinepool.go]: Yup.

File: docs/user/openstack/customization.md [hunk 5]
```diff
@@ -16,6 +16,9 @@ Beyond the [platform-agnostic `install-config.yaml` properties](../customization
 ## Machine pools
 
 * `type` (optional string): The OpenStack flavor name for machines in the pool.
+* `rootVolume` (optional object): Defines the root volume for instances in the machine pool. The instances use ephemeral disks if not set.
+  * `size` (integer): Size of the root volume in GB.
+  * `type` (string): The volume pool to create the volume from.
```
Comment (wking) [inline docs/user/openstack/customization.md]: These should be "optional integer" and "optional string" (or maybe "required string"?) respectively.  Or maybe they're both required?  Either way, say "optional" or "required", and if optional, say something about the defaults, unless the default is "we'll magically pick something that works".

File: pkg/types/openstack/machinepool.go [hunk 6]
```diff
@@ -6,6 +6,9 @@ type MachinePool struct {
 	// FlavorName defines the OpenStack Nova flavor.
 	// eg. m1.large
 	FlavorName string `json:"type"`
+
+	// RootVolume defines the root volume for instances in the machine pool.
+	RootVolume *RootVolume `json:"rootVolume,omitempty"`
```
Comment (abhinavdahiya) [inline pkg/types/openstack/machinepool.go]: needs the information about `if optional` and what is the default when nil

File: pkg/types/openstack/validation/machinepool.go [hunk 7]
```diff
@@ -8,5 +8,17 @@ import (
 
 // ValidateMachinePool checks that the specified machine pool is valid.
 func ValidateMachinePool(p *openstack.MachinePool, fldPath *field.Path) field.ErrorList {
-	return field.ErrorList{}
+	allErrs := field.ErrorList{}
+
+	// Validate Volumes
+	if p.RootVolume != nil {
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/machinepool.go]: based on https://github.com/openshift/installer/pull/2394/files#diff-f4a1bfbddde14b37dea44f2357757492R35-R40

both the values must be set when RootVolume is not-nil, so why do we have 
`p.RootVolume.Type == "" && p.RootVolume.Size > 0`
`p.RootVolume.Type != "" && p.RootVolume.Size <= 0`

why not 2 conditions 
`p.RootVolume.Type == ""` -> error
`p.RootVolume.Size <=0` -> error

File: pkg/types/openstack/validation/machinepool.go [hunk 8]
```diff
@@ -8,5 +8,17 @@ import (
...
+		if p.RootVolume.Type == "" && p.RootVolume.Size > 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("Type"), p.RootVolume.Type, "Volume type must be specified to use root volumes"))
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/machinepool.go]: `fldPath.Child("rootVoulme").Child("type")`

File: pkg/types/openstack/validation/machinepool.go [hunk 9]
```diff
@@ -8,5 +8,17 @@ import (
...
+		}
+		if p.RootVolume.Type != "" && p.RootVolume.Size <= 0 {
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("Size"), p.RootVolume.Type, "Volume size must be greater than zero to use root volumes"))
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/machinepool.go]: `fldPath.Child("rootVolume").Child("size")`

File: pkg/types/openstack/machinepool.go [hunk 10]
```diff
@@ -17,4 +20,21 @@ func (o *MachinePool) Set(required *MachinePool) {
 	if required.FlavorName != "" {
 		o.FlavorName = required.FlavorName
 	}
+
+	if required.RootVolume != nil {
+		if required.RootVolume.Size != 0 {
+			o.RootVolume.Size = required.RootVolume.Size
+		}
+		if required.RootVolume.Type != "" {
```
Comment (wking) [inline pkg/types/openstack/machinepool.go]: If valid `MachinePool` set both `Size` and `Type`, it is surprising to me that we have two separate conditionals here.  Can we validate `required` and then just use:

```golang
if required.RootVolume.Size != 0 {
  o.RootVolume.Size = required.RootVolume.Size
  o.RootVolume.Type = required.RootVolume.Type
}
```

Do we actually use this `Set` call?  If so, maybe `type` and `size` actually _are_ optional as long as someone in this chain defines them?

File: pkg/types/openstack/machinepool.go (hunk 6 above)
Comment (iamemilio) [reply pkg/types/openstack/machinepool.go]: this is unset when nil, can I just leave it at that

Comment (abhinavdahiya) [reply pkg/types/openstack/machinepool.go]: > https://github.com/openshift/installer/pull/2394/files#diff-e32975a988b75f5459f56a7c4e5affe7R19
`The instances use ephemeral disks if not set.` should be enough.

File: pkg/types/openstack/validation/machinepool.go [hunk 11]
```diff
@@ -7,6 +7,26 @@ import (
 )
 
 // ValidateMachinePool checks that the specified machine pool is valid.
-func ValidateMachinePool(p *openstack.MachinePool, fldPath *field.Path) field.ErrorList {
-	return field.ErrorList{}
+func ValidateMachinePool(platform *openstack.Platform, p *openstack.MachinePool, fldPath *field.Path) field.ErrorList {
+	allErrs := field.ErrorList{}
+
+	// Validate Root Volumes
```
Comment (abhinavdahiya) [inline pkg/types/openstack/validation/machinepool.go]: I don't understand this change? Why are we setting when validating? That's not how we do it for other platforms...

File: pkg/types/openstack/machinepool.go (hunk 6 above)
Comment (mandre) [reply pkg/types/openstack/machinepool.go]: Done

File: pkg/types/openstack/validation/machinepool.go (hunk 7 above)
Comment (mandre) [reply pkg/types/openstack/validation/machinepool.go]: Done

**Comment by mandre** [comment]
@abhinavdahiya @wking Thanks for the reviews. I believe we've addressed all the comments, could you take a look again when you're around?

**Pull request merged**